### ISSUE TITLE
ORCA fails to build a UnionAll plan

### ISSUE THREAD

**Issue by darthunix** [greenplum-db/gpdb#11276]
### Greenplum version or build

Tested and reproduced on Orca from `master` and `6X_STABLE` branches

### Expected behavior

Build and normalize a correct union all histogram for a query
```sql
explain
select * from "reporting"."data_mart"."unrealized_pnl_v8_marginal_forex_opened_stg" where open_positions_date = '2020-09-29'::date
UNION ALL
select * from "reporting"."data_mart"."unrealized_pnl_v8_marginal_forex_closed_stg" where open_positions_date = '2020-09-29'::date;

```

### Actual behavior

Fails on asserts while building a UnionAll histogram from united histograms' buckets. If been built without asserts utilizes all memory + CPU core and fails with OOM.
```
* thread #1, queue = 'com.apple.main-thread', stop reason = User Expression thread plan
    frame #0: 0x0000000102ae8186 libgpos.dylib`gpos::CException::Raise(filename="../libnaucrates/src/statistics/CHistogram.cpp", line=1643, major=1, minor=2) at CException.cpp:103:17
  * frame #1: 0x00000001026ec209 libnaucrates.dylib`gpnaucrates::CHistogram::CombineBuckets(mp=0x00007f940d217760, buckets=0x00007f93fceeecc0, desired_num_buckets=299))>*, unsigned int) at CHistogram.cpp:1643:2
    frame #2: 0x00000001026eb95d libnaucrates.dylib`gpnaucrates::CHistogram::MakeUnionAllHistogramNormalize(this=0x00007f9417914430, rows=(m_d = 1079.725248), histogram=0x00007f93fcdbfe00, rows_other=(m_d = 1713.7444499999997)) const at CHistogram.cpp:1530:3
    frame #3: 0x0000000102716010 libnaucrates.dylib`gpnaucrates::CUnionAllStatsProcessor::CreateStatsForUnionAll(mp=0x00007f940d217760, stats_first_child=0x00007f9419bdd2e0, stats_second_child=0x00007f93ff249640, output_colids=0x00007f93fe2b6960, first_child_colids=0x00007f93ff273f40, second_child_colids=0x00007f93ff27a4a0))>*, gpos::CDynamicPtrArray<unsigned int, &(void gpos::CleanupDelete<unsigned int>(unsigned int*))>*, gpos::CDynamicPtrArray<unsigned int, &(void gpos::CleanupDelete<unsigned int>(unsigned int*))>*) at CUnionAllStatsProcessor.cpp:67:29
    frame #4: 0x0000000104f7a83c libgpopt.dylib`gpopt::CLogicalUnionAll::PstatsDeriveUnionAll(mp=0x00007f940d217760, exprhdl=0x00007ffeed8c42f8) at CLogicalUnionAll.cpp:170:24
    frame #5: 0x0000000104f7a908 libgpopt.dylib`gpopt::CLogicalUnionAll::PstatsDerive(this=0x00007f9414913490, mp=0x00007f940d217760, exprhdl=0x00007ffeed8c42f8, (null)=0x00007f94179c1310))>*) const at CLogicalUnionAll.cpp:198:9
    frame #6: 0x0000000104f2d1ad libgpopt.dylib`gpopt::CExpressionHandle::DeriveRootStats(this=0x00007ffeed8c42f8, stats_ctxt=0x00007f94179c1310))>*) at CExpressionHandle.cpp:403:28
    frame #7: 0x0000000104f2d900 libgpopt.dylib`gpopt::CExpressionHandle::DeriveStats(this=0x00007ffeed8c42f8, stats_ctxt=0x00007f94179c1310, fComputeRootStats=true))>*, bool) at CExpressionHandle.cpp:478:3
    frame #8: 0x00000001050000a8 libgpopt.dylib`gpopt::CGroupExpression::PstatsRecursiveDerive(this=0x00007f94150e0630, (null)=0x00007f940d217760, pmpGlobal=0x00007f940d217760, prprel=0x00007f94179c11a0, stats_ctxt=0x00007f94179c1310, fComputeRootStats=true))>*, bool) at CGroupExpression.cpp:1006:10
    frame #9: 0x0000000104fef3a3 libgpopt.dylib`gpopt::CGroup::PstatsRecursiveDerive(this=0x00007f941205ea40, pmpLocal=0x00007f940d217760, pmpGlobal=0x00007f940d217760, prprel=0x00007f94179c11a0, stats_ctxt=0x00007f94179c1310))>*) at CGroup.cpp:1505:22
    frame #10: 0x0000000104f2e90d libgpopt.dylib`gpopt::CExpressionHandle::DeriveStats(this=0x00007ffeed8c4530, pmpLocal=0x00007f940d217760, pmpGlobal=0x00007f940d217760, prprel=0x0000000000000000, stats_ctxt=0x0000000000000000))>*) at CExpressionHandle.cpp:648:30
    frame #11: 0x0000000104ee48be libgpopt.dylib`gpopt::CEngine::DeriveStats(pmpLocal=0x00007f940d217760, pmpGlobal=0x00007f940d217760, pgroup=0x00007f941205ea40, prprel=0x0000000000000000) at CEngine.cpp:462:10
    frame #12: 0x0000000104ee477f libgpopt.dylib`gpopt::CEngine::DeriveStats(this=0x00007ffeed8c6710, pmpLocal=0x00007f940d217760) at CEngine.cpp:441:3
    frame #13: 0x0000000104eea5c1 libgpopt.dylib`gpopt::CEngine::FinalizeExploration(this=0x00007ffeed8c6710) at CEngine.cpp:1456:3
    frame #14: 0x000000010500f9b2 libgpopt.dylib`gpopt::CJobGroupExploration::EevtExploreChildren(psc=0x00007ffeed8c6508, pjOwner=0x00007f93fbcfd2f8) at CJobGroupExploration.cpp:232:17
    frame #15: 0x000000010500ff80 libgpopt.dylib`gpopt::CJobStateMachine<gpopt::CJobGroupExploration::EState, (gpopt::CJobGroupExploration::EState)3, gpopt::CJobGroupExploration::EEvent, (gpopt::CJobGroupExploration::EEvent)3>::FRun(this=0x00007f93fbcfd368, psc=0x00007ffeed8c6508, pjOwner=0x00007f93fbcfd2f8) at CJobStateMachine.h:162:21
    frame #16: 0x000000010500fdb0 libgpopt.dylib`gpopt::CJobGroupExploration::FExecute(this=0x00007f93fbcfd2f8, psc=0x00007ffeed8c6508) at CJobGroupExploration.cpp:253:15
    frame #17: 0x000000010502bda0 libgpopt.dylib`gpopt::CScheduler::FExecute(this=0x00007ffeed8c6538, pj=0x00007f93fbcfd2f8, psc=0x00007ffeed8c6508) at CScheduler.cpp:315:22
    frame #18: 0x000000010502b75e libgpopt.dylib`gpopt::CScheduler::ExecuteJobs(this=0x00007ffeed8c6538, psc=0x00007ffeed8c6508) at CScheduler.cpp:123:21
    frame #19: 0x000000010502b6d9 libgpopt.dylib`gpopt::CScheduler::Run(pv=0x00007ffeed8c6508) at CScheduler.cpp:95:17
    frame #20: 0x0000000104eecdcf libgpopt.dylib`gpopt::CEngine::Optimize(this=0x00007ffeed8c6710) at CEngine.cpp:1717:3
    frame #21: 0x0000000104fe0e70 libgpopt.dylib`gpopt::COptimizer::PexprOptimize(mp=0x00007f940d217760, pqc=0x00007f94141cca50, search_stage_array=0x0000000000000000))>*) at COptimizer.cpp:422:6
    frame #22: 0x0000000104fe046c libgpopt.dylib`gpopt::COptimizer::PdxlnOptimize(mp=0x00007f940d217760, md_accessor=0x00007f940e3ec250, query=0x00007f941620fe90, query_output_dxlnode_array=0x00007f9415fce860, cte_producers=0x00007f94165a86d0, pceeval=0x0000000000000000, ulHosts=52, ulSessionId=1, ulCmdId=1, search_stage_array=0x0000000000000000, optimizer_config=0x00007f940d23fe40, szMinidumpFileName="/Users/darthunix/Downloads/ADBDEV-1181.mdp"))> const*, gpos::CDynamicPtrArray<gpdxl::CDXLNode, &(void gpos::CleanupRelease<gpdxl::CDXLNode>(gpdxl::CDXLNode*))> const*, gpopt::IConstExprEvaluator*, unsigned int, unsigned int, unsigned int, gpos::CDynamicPtrArray<gpopt::CSearchStage, &(void gpos::CleanupDelete<gpopt::CSearchStage>(gpdxl::CDXLNode*))>*, gpopt::COptimizerConfig*, char const*) at COptimizer.cpp:318:29
    frame #23: 0x0000000104f1a9ff libgpopt.dylib`gpopt::CMinidumperUtils::PdxlnExecuteMinidump(mp=0x00007f940d217760, md_accessor=0x00007f940e3ec250, pdxlmd=0x00007f9415f3aad0, file_name="/Users/darthunix/Downloads/ADBDEV-1181.mdp", ulSegments=52, ulSessionId=1, ulCmdId=1, optimizer_config=0x00007f940d23fe40, pceeval=0x0000000000000000) at CMinidumperUtils.cpp:334:15
    frame #24: 0x0000000104f1a562 libgpopt.dylib`gpopt::CMinidumperUtils::PdxlnExecuteMinidump(mp=0x00007f940d217760, pdxlmd=0x00007f9415f3aad0, file_name="/Users/darthunix/Downloads/ADBDEV-1181.mdp", ulSegments=52, ulSessionId=1, ulCmdId=1, optimizer_config=0x00007f940d23fe40, pceeval=0x0000000000000000) at CMinidumperUtils.cpp:282:21
    frame #25: 0x0000000104f1a3f2 libgpopt.dylib`gpopt::CMinidumperUtils::PdxlnExecuteMinidump(mp=0x00007f940d217760, file_name="/Users/darthunix/Downloads/ADBDEV-1181.mdp", ulSegments=52, ulSessionId=1, ulCmdId=1, optimizer_config=0x00007f940d23fe40, pceeval=0x0000000000000000) at CMinidumperUtils.cpp:250:3
    frame #26: 0x000000010233ea2d gporca_test`PvExec(pv=0x00007ffeed8c8508) at main.cpp:387:25
    frame #27: 0x0000000102b04180 libgpos.dylib`gpos::CTask::Execute(this=0x00007f940d1a22f0) at CTask.cpp:129:12
    frame #28: 0x0000000102b077ba libgpos.dylib`gpos::CWorker::Execute(this=0x00007ffeed8c8420, task=0x00007f940d1a22f0) at CWorker.cpp:80:11
    frame #29: 0x0000000102b029d2 libgpos.dylib`gpos::CAutoTaskProxy::Execute(this=0x00007ffeed8c83d8, task=0x00007f940d1a22f0) at CAutoTaskProxy.cpp:285:11
    frame #30: 0x0000000102ad8d52 libgpos.dylib`::gpos_exec(params=0x00007ffeed8c84c0) at _api.cpp:246:9
    frame #31: 0x000000010233e5ee gporca_test`main(iArgs=3, rgszArgs=0x00007ffeed8c8570) at main.cpp:442:6
    frame #32: 0x00007fff20333631 libdyld.dylib`start + 1s
```

### Step to reproduce the behavior

1. Download [minidump](https://drive.google.com/file/d/1Oyy0xYpvPE3-fccNiqgdG6mIQHiA7-F2/view?usp=sharing) file (it has only DXL query and statistics, plan fails to build)
1. `./server/gporca_test -d ADBDEV-1181.mdp`

### Additional information

Technically we fail an assert `sum(bucket frequency) <= 1 + epsilon` when we combine buckets from histo#1 and hito#2 into a new union all histogram for a column `user_id`. But the root of the problem seems to be deeper - histo#1 and histo#2 are not normalized properly and overestimate `1 + epsilon` as well.

I've collected their bucket information with lldb and its python scripting (you can get this data yourself if you like, but I attach result CSV files) and original postgres statistics.
1. Histogram#1 ([this.csv](https://drive.google.com/file/d/1fljFRkdnN5sQ7aKbHt6qu5PwHbe12XUP/view?usp=sharing)) - frame 2:
```python
thread = lldb.process.selected_thread
suitable_frames = [f for f in thread.frames if "MakeUnionAllHistogramNormalize" in f.name]
frame = suitable_frames[0]
size = frame.FindVariable("this.csv").GetChildMemberWithName("m_histogram_buckets").GetChildMemberWithName("m_size").GetValueAsUnsigned()
result = ""
for i in range(0, size):
    is_lower_closed = frame.EvaluateExpression("this->m_histogram_buckets->m_elems[%s]->IsLowerClosed()" % i).GetValue()
    is_upper_closed = frame.EvaluateExpression("this->m_histogram_buckets->m_elems[%s]->IsUpperClosed()" % i).GetValue()
    lb = frame.EvaluateExpression("this->m_histogram_buckets->m_elems[%s]->GetLowerBound()->m_datum->GetValAsDouble().m_d" % i).GetValue()
    ub = frame.EvaluateExpression("this->m_histogram_buckets->m_elems[%s]->GetUpperBound()->m_datum->GetValAsDouble().m_d" % i).GetValue()
    freq = frame.EvaluateExpression("this->m_histogram_buckets->m_elems[%s]->GetFrequency().m_d" % i).GetValue()
    ndv = frame.EvaluateExpression("this->m_histogram_buckets->m_elems[%s]->GetNumDistinct().m_d" % i).GetValue()
    result += "[" if is_lower_closed == 'true' else "("
    result += lb + "," + ub
    result += "]|" if is_upper_closed == 'true' else ")|"
    result += freq + "|" + ndv
    result += "\n"

with open('this.csv', 'w') as f:
    f.write(result)
```
2. Histogram#2 ([histo.csv](https://drive.google.com/file/d/1EMIZpyTpQL9AL5kzLl2KMjdA0yqLWxWI/view?usp=sharing)) - frame 2:
```python
thread = lldb.process.selected_thread
suitable_frames = [f for f in thread.frames if "MakeUnionAllHistogramNormalize" in f.name]
frame = suitable_frames[0]
size = frame.FindVariable("histogram").GetChildMemberWithName("m_histogram_buckets").GetChildMemberWithName("m_size").GetValueAsUnsigned()
result = ""
for i in range(0, size):
    is_lower_closed = frame.EvaluateExpression("histogram->m_histogram_buckets->m_elems[%s]->IsLowerClosed()" % i).GetValue()
    is_upper_closed = frame.EvaluateExpression("histogram->m_histogram_buckets->m_elems[%s]->IsUpperClosed()" % i).GetValue()
    lb = frame.EvaluateExpression("histogram->m_histogram_buckets->m_elems[%s]->GetLowerBound()->m_datum->GetValAsDouble().m_d" % i).GetValue()
    ub = frame.EvaluateExpression("histogram->m_histogram_buckets->m_elems[%s]->GetUpperBound()->m_datum->GetValAsDouble().m_d" % i).GetValue()
    freq = frame.EvaluateExpression("histogram->m_histogram_buckets->m_elems[%s]->GetFrequency().m_d" % i).GetValue()
    ndv = frame.EvaluateExpression("histogram->m_histogram_buckets->m_elems[%s]->GetNumDistinct().m_d" % i).GetValue()
    result += "[" if is_lower_closed == 'true' else "("
    result += lb + "," + ub
    result += "]|" if is_upper_closed == 'true' else ")|"
    result += freq + "|" + ndv
    result += "\n"

with open('histo.csv', 'w') as f:
    f.write(result)
```
3. Union all histogram ([buckets.csv](https://drive.google.com/file/d/1l0KVnwZRRfqMeYh6mqwZvNEuXsQ7Op0A/view?usp=sharing)) - frame 1:
```python
thread = lldb.process.selected_thread
suitable_frames = [f for f in thread.frames if "CombineBuckets" in f.name]
frame = suitable_frames[0]
size = frame.FindVariable("buckets").GetChildMemberWithName("m_size").GetValueAsUnsigned()
result = ""
for i in range(0, size):
    is_lower_closed = frame.EvaluateExpression("buckets->m_elems[%s]->IsLowerClosed()" % i).GetValue()
    is_upper_closed = frame.EvaluateExpression("buckets->m_elems[%s]->IsUpperClosed()" % i).GetValue()
    lb = frame.EvaluateExpression("buckets->m_elems[%s]->GetLowerBound()->m_datum->GetValAsDouble().m_d" % i).GetValue()
    ub = frame.EvaluateExpression("buckets->m_elems[%s]->GetUpperBound()->m_datum->GetValAsDouble().m_d" % i).GetValue()
    freq = frame.EvaluateExpression("buckets->m_elems[%s]->GetFrequency().m_d" % i).GetValue()
    ndv = frame.EvaluateExpression("buckets->m_elems[%s]->GetNumDistinct().m_d" % i).GetValue()
    result += "[" if is_lower_closed == 'true' else "("
    result += lb + "," + ub
    result += "]|" if is_upper_closed == 'true' else ")|"
    result += freq + "|" + ndv
    result += "\n"

with open('buckets.csv', 'w') as f:
    f.write(result)
```
If we upload all these histograms to Greenplum and count sum frequency for all each histogram we'll get `1.000017` for histo#1, `1.000023` for histo#2 and `1.000021` for resulting union all histogram. It is bigger than `1 + epsilon=0.00001`.
```sql
create table histo(bounds numrange, freq numeric, ndv numeric)
distributed randomly;
copy histo from '/Users/darthunix/Downloads/histo.csv' csv delimiter '|';

create table buckets(bounds numrange, freq numeric, ndv numeric)
distributed randomly;
copy buckets from '/Users/darthunix/Downloads/buckets.csv' csv delimiter '|';

create table this(bounds numrange, freq numeric, ndv numeric)
distributed randomly;
copy this from '/Users/darthunix/Downloads/this.csv' csv delimiter '|';

select sum(freq) from this;                                                                                                                                                                           
           sum                                                                                                                                                                                              
--------------------------                                                                                                                                                                                  
 1.0000169999999999849939                                                                                                                                                                                   
(1 row)                                                                                                                                                                                                     
                                                                                                                                                                                                            
select sum(freq) from histo;                                                                                                                                                                          
           sum                                                                                                                                                                                              
--------------------------                                                                                                                                                                                  
 1.0000230000000000103208                                                                                                                                                                                   
(1 row)                                                                                                                                                                                                                                                                                                                                                                                                              

select sum(freq) from buckets;                                                                   
           sum            
--------------------------
 1.0000206808943039926027
(1 row)
```
I suppose that this histograms were already improperly normalized when we added singletons to postgres histograms. I have also collected original postgres [statistics](https://drive.google.com/file/d/1gXM6Ufj9Erx-aV39EqPC0Que_mzlwzmz/view?usp=sharing). It can be loaded into postgres with
```sql
create table stat_buff (
    starelid oid,
    staattnum smallint,
    stainherit boolean,
    stanullfrac real,
    stawidth int,
    stadistinct real,
    stakind1 smallint,
    stakind2 smallint,
    stakind3 smallint,
    stakind4 smallint,
    stakind5 smallint,
    staop1 oid,
    staop2 oid,
    staop3 oid,
    staop4 oid,
    staop5 oid,
    stanumbers1 real[],
    stanumbers2 real[],
    stanumbers3 real[],
    stanumbers4 real[],
    stanumbers5 real[],
    stavalues1 text,
    stavalues2 text,
    stavalues3 text,
    stavalues4 text,
    stavalues5 text
)
distributed randomly;

copy stat_buff from '/tmp/stats.txt' csv;
```
To retreive `user_id` (attribute 3) stats in a manner `pg_statistics` does run
```sql
select
    case
        when starelid = 28045475 then 'data_mart.unrealized_pnl_v8_marginal_forex_closed_stg'
        else 'data_mart.unrealized_pnl_v8_marginal_forex_opened_stg'
    end as relname,
    stainherit AS inherited,
    stanullfrac AS null_frac,
    stawidth AS avg_width,
    stadistinct AS n_distinct,
    stavalues1::int[] as most_common_vals,
    stanumbers1::numeric[] as most_common_freqs,
    stavalues2::int[] as histogram_bounds
from stat_buff where staattnum = 3;
```

At this moment I've stuck as I can't find an algorithm in the code where we transform postgres statistics to Orca one and add MCV as singleton buckets to original histograms. I need your advice here.

### A possibly wrong assert

Also I've found that there is a possibly wrong [IsValid()](https://github.com/greenplum-db/gpdb/blob/2b61d021f4f872e16e112e48f5aa70630263b575/src/backend/gporca/libnaucrates/src/statistics/CHistogram.cpp#L576) assert (seems that a round bracket is placed in a wrong place) - but it doesn't affect current problem:
```
 	// At most, we would be overestimating by 0.5 freq per bucket, so give (num_buckets / 2)
 	// We also give an additional bit for internal rounding within ORCA calculations.
 	if (GetFrequency() >
-		CDouble(1.0) + (CStatistics::Epsilon * GetNumBuckets() / 2 + 1))
+		CDouble(1.0) + CStatistics::Epsilon * (GetNumBuckets() / 2 + 1))
 	{
 		return false;
 	}
```

**Comment by vraghavan78** [comment]
@chrishajas can u please check on this when u back in Jan?

**Comment by chrishajas** [comment]
@darthunix I tried the minidump on the latest 6X and it works on the retail build. This appears identical to a recent bug that was fixed in https://github.com/greenplum-db/gpdb/pull/10994, where queries involving union/union all were not completing and then going OOM.

The assertion failure seems to be somewhat unrelated (though a legitimate issue), as it deals with frequencies rather than the boundary conditions that we use when combining buckets.

Can you check whether this issue still exists after that commit?

**Comment by darthunix** [comment]
@chrishajas thanks, you are correct - https://github.com/greenplum-db/gpdb/pull/10994 fixed the problem and since `6.12.0` orca is not affected by this bug. This issue should be closed.
